### ISSUE TITLE
cloud-config in vendordata overriden by user-data

### ISSUE THREAD

**Issue by ubuntu-server-builder** [canonical/cloud-init#3411]
This bug was originally filed in Launchpad as [LP: #1837530](https://bugs.launchpad.net/cloud-init/+bug/1837530)
<details>
<summary>Launchpad details</summary>
<pre>
affected_projects = []
assignee = None
assignee_name = None
date_closed = 2019-09-22T04:17:36.749205+00:00
date_created = 2019-07-23T09:15:39.460975+00:00
date_fix_committed = None
date_fix_released = None
id = 1837530
importance = undecided
is_complete = True
lp_url = https://bugs.launchpad.net/cloud-init/+bug/1837530
milestone = None
owner = ameretat-reith
owner_name = r
private = False
status = expired
submitter = ameretat-reith
submitter_name = r
tags = []
duplicates = []
</pre>
</details>

_Launchpad user **r(ameretat-reith)** wrote on 2019-07-23T09:15:39.460975+00:00_

This may be a bug or valid behaviour, but by reading little available vendordata documentations, I think either documentation is invalid or at-least lacking something.

I'm trying to boot instances using config-drive and openstack dialect. There is no Nova or metadata service involved, just plain config-drive which happens to have user-data, vendor-data, network-data and metadata in following hierarchy:

─── openstack
    ├── content
    │   └── hotplug-cpu-udev.rules
    └── latest
        ├── meta_data.json
        ├── network_data.json
        ├── user_data
        └── vendor_data.json

I'm using vendor_data to supply general instructions which is shared between every machine. It is provided by me; as infrastructure provider not user and contains instructions to make machines works with platform that I provide. So I think it's valid to have these general instructions in vendor_data and let users configure their machines as their wish by providing user_data.

The problem is whenever user_data exists, vendor_data will be erased. Actually logs says it's processed and there will be /var/lib/cloud/instance/vendor-cloud-config.txt but If user_data and vendor_data both have `runcmd` for example, /var/lib/cloud/instance/scripts/runcmd will contain just user_data scripts.

user_data cloud-config is like:

#cloud-config
hostname: ubuntu.default
manage_etc_hosts: localhost
growpart:
  mode: auto
  devices:
  - /
  ignore_growroot_disabled: false
runcmd:
- wget http://example/img -O /tmp/img
resize_rootfs: true
ssh_pwauth: false

So my understanding is, It should not prevent vendor-data to be ran. For the reference vendor_data.json content is:

{"cloud-init": "#cloud-config\nruncmd:\n- touch /tmp/vendor"}

So, should I change user provided user_data to allow vendor data execution? That would be a little weird since documentation said the other way; user_data should contain instructions to disallow vendordata executions. And if documentation is wrong, it's sill doesn't look nice to ask users change their cloud-config to be run properly on my platform.

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **r(ameretat-reith)** wrote on 2019-07-23T09:22:39.425116+00:00_

I forgot to mention cloud-init version which is 18.4

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Ryan Harper(raharper)** wrote on 2019-07-23T14:05:40.936390+00:00_

Thanks for filing the bug.

You're correct that user-data will override vendor-data where the keys match.  This is by design as cloud-init wants to allow the users to fully control their instance.

In your example, you're providing a vendor-data config which uses 'runcmd'; and the user-data also supplies this key as well.

The cloud-init merging will have user-data overwrite the vendor-data.

https://cloudinit.readthedocs.io/en/latest/topics/merging.html


Further, the documentation on vendor-data appears to be quite clear about behavior:

https://cloudinit.readthedocs.io/en/latest/topics/vendordata.html

Do you have any suggestions to make things more clear?

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **r(ameretat-reith)** wrote on 2019-07-23T16:14:23.849100+00:00_

Thanks fro replying Ryan,

You are correct and documentations correctly described behaviour. I thought merging happens only if cloud-init config is #cloud-config-jsonp part, maybe that paragraph could be rephrased a little... Also it would help if there was a link to Merging document.. I also was mistaken that I can use #cloud-config-jsonp in vendor_data not in user_data and when cloud-init stuck in my miss-configuration, I used cloud-config which caused merging to override vendor_data. I also spent some time to figure out how should I push vendor_data cloud-config in config-drive. I didn't found it documented in cloud-init or openstack configurations and finally understood from code. (it should be value of `cloud-init` key in vendor_data.json which was not clear to me)

I doubt a little about usefulness of merging in this scenario... One point is many modules logic can be applied by `runcmd`, so in user's point of view, it may doesn't matter to discard one vendor_data module and accept one another, because actions doesn't match well with modules and user concerns about actions. I think user probably just wants disable vendor_data or let it run. I also found it not very nice to ask users to change their cloud-init because I have some vendor_data for my special platfrom... They probably already have their cloud-config and want bring it to my infrastructure.
